### ISSUE TITLE
False positives when deffering import of typing special forms

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#9664]
**Describe the bug**
Deferring an import changes the behavior of the typing special form

**Code or Screenshots**
```py
# _typings.py
from __future__ import annotations

TYPE_CHECKING = False

if TYPE_CHECKING:
    from typing import Any, Final, Literal, Self, final
else:
    def final(f):  # noqa: ANN001
        return f


def __getattr__(name: str):
    if name == "final":  # this one actually executes at runtime
        return final

    if name in {"Any", "Final", "Literal", "Self"}:
        import typing
        return getattr(typing, name)

    msg = f"module {__name__!r} has no attribute {name!r}"
    raise AttributeError(msg)


__all__ = ["Any", "Final", "Literal", "Self", "final"]
```

```py
# other module
from ._typings import Any, Final, Literal

type HandleableSignals = Literal["SIGINT", "SIGTERM", "SIGBREAK", "SIGHUP"]
type SignalTuple = tuple[HandleableSignals, ...]

default_handled: Final = "SIGINT", "SIGTERM", "SIGBREAK"
```

```
"Final" is not allowed in this context 
Type "tuple[Literal['SIGINT'], Literal['SIGTERM'], Literal['SIGBREAK']]" is not assignable to declared type "Final"
  "tuple[Literal['SIGINT'], Literal['SIGTERM'], Literal['SIGBREAK']]" is not assignable to "Final"
```

This works fine when importing directly from typing

**VS Code extension or command-line**
via Pylance 2024.12.1

**Comment by erictraut** [comment]
Yes, this is a known (and as-designed) limitation of pyright. The `Final` symbol must be imported directly from `typing` and not re-exported from other modules if you want it to be treated as `Final`. In most programming languages, this would be a keyword rather than a special form symbol. Pyright needs to understand `Final` in the "binding" phase when it's building symbol tables. This is before it's able to do semantic type analysis. 

I'm going to close this because pyright is working as intended here. If you want your code to work with pyright, you'll need to import `Final` directly from `typing`.
